client/signup/steps/p2-complete-profile/index.jsx

```diff
@@ -0,0 +1,260 @@
+import ValidationFieldset from 'calypso/signup/validation-fieldset';
+import { saveSignupStep, submitSignupStep } from 'calypso/state/signup/progress/actions';
+import { saveUserSettings } from 'calypso/state/user-settings/actions';
+import './style.scss';
+
+/**
+ * Constants
+ */
+const VALIDATION_DELAY_AFTER_FIELD_CHANGES = 1500;
+const ERROR_CODE_MISSING_FULL_NAME = 123; // Random number, we don't need it.
+const ERROR_CODE_FROM_LOCAL_STORAGE = 7331; // Random number, we don't need it.
+
+class P2CompleteProfile extends Component {
+	constructor( props ) {
+		super( props );
+
+		let initialState;
+
+		if ( props?.step?.form ) {
+			initialState = props.step.form;
+
+			if ( ! isEmpty( props.step.errors ) ) {
+				const errorMessage = props.step.errors[ 0 ].message;
+
+				this.logValidationErrorToLogstash( ERROR_CODE_FROM_LOCAL_STORAGE, errorMessage );
+
+				initialState = formState.setFieldErrors(
+					formState.setFieldsValidating( initialState ),
+					{
+						site: {
+							[ ERROR_CODE_FROM_LOCAL_STORAGE ]: errorMessage,
+						},
+					},
+					true
+				);
+			}
+		} else {
+			initialState = {
+				fullName: {
+					value: '',
+				},
+			};
+		}
+
+		this.formStateController = new formState.Controller( {
+			fieldNames: [ 'fullName' ],
+			sanitizerFunction: this.sanitize,
+			validatorFunction: this.validate,
+			onNewState: ( state ) => {
+				this.setState( { form: state } );
+			},
+			onError: this.handleFormControllerError,
+			debounceWait: VALIDATION_DELAY_AFTER_FIELD_CHANGES,
+			hideFieldErrorsOnChange: true,
+			initialState: initialState,
+			skipSanitizeAndValidateOnFieldChange: true,
+		} );
+
+		this.state = {
+			form: this.formStateController.getInitialState(),
+			isSubmitting: false,
+			suggestedSubdomains: [],
+			lastInvalidSite: '',
+		};
+	}
+
+	componentWillUnmount() {
+		this.save();
+	}
+
+	validate = ( fields, onComplete ) => {
+		const messages = {};
+
+		if ( isEmpty( fields.fullName ) || fields.fullName.length < 3 ) {
+			messages.fullName = {
+				[ ERROR_CODE_MISSING_FULL_NAME ]: this.props.translate(
+					'Please enter your full name (3 characters or more).'
+				),
+			};
+		}
+
+		onComplete( null, messages );
+	};
+
+	handleSubmit = ( event ) => {
+		event.preventDefault();
+
+		this.setState( { isSubmitting: true } );
+
+		this.formStateController.handleSubmit( ( hasErrors ) => {
+			const fullName = formState.getFieldValue( this.state.form, 'fullName' );
+
+			if ( hasErrors ) {
+				this.setState( { isSubmitting: false } );
+
+				return;
+			}
+
+			recordTracksEvent( 'calypso_signup_p2_complete_profile_step_submit' );
+
+			// API calls
+
+			this.props.saveUserSettings( { display_name: fullName } );
+
+			const stepData = {
+				stepName: this.props.stepName,
+				form: this.state.form,
+				fullName,
+			};
+
+			this.props.submitSignupStep( stepData );
+
+			this.props.goToNextStep();
+		} );
+	};
+
+	handleFormControllerError = ( error ) => {
```

Comment:
Abstractions like that are why I really recommend not using `formState`. Just for the protocol 😉